### PR TITLE
holochain-go: init at 0.1.0-alpha

### PR DESCRIPTION
**Pull request by gavinrogers**
Holochain is a platform for creating an unenclosable carrier upon which
highly-scalable dApps and smart contracts can be built. This is the holochain
team's initial implimentation in go. A rust version is under development.

###### Motivation for this change

Add the holochain core to nixpkgs to allow app developers to bootstrap more quickly: allow people to run their own holochain node.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"` (nothing depends on this change)
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after) 
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by gavinrogers** [comment]
@zimbatm How does this look to you?

**Comment by gavinrogers** [comment]
@edolstra this is what I was telling you about last week

File: pkgs/servers/holochain-go/default.nix [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  goDeps = ./deps.nix;
+
+  meta = {
+    description = "core implementation of validating DHT dApps in go";
+    longDescription = "Holographic storage for distributed applications -- a validating monotonic DHT backed by authoritative hashchains for data provenance";
+    homepage = "https://holochain.org/";
+    downloadPage = "https://developer.holochain.org/";
+    license = stdenv.lib.licenses.gpl3;
+    maintainers = with stdenv.lib.maintainers; [ gavin ];
+    platforms = stdenv.lib.platforms.all;
```
Comment (zimbatm) [inline pkgs/servers/holochain-go/default.nix]: if `platforms` is not set it will default to `go.meta.platforms` which is probably what you want.

**Review by zimbatm — APPROVED** [review]
looking good!

**Comment by zimbatm** [comment]
@GrahamcOfBorg holochain-go

**Comment by xeji** [comment]
@GrahamcOfBorg build holochain-go

**Comment by zimbatm** [comment]
@gavinrogers has holochain-go been exclusively developed on a mac? The Linux compilation is failing with an exit status 8.

**Comment by gavinrogers** [comment]
@zimbatm No it's been developed on linux machines as well. It's complaining about github.com/gxed/eventfd ... which isn't called by holochain-proto (that I can tell). Is it being called by a dependency? I'll add it

File: pkgs/servers/holochain-go/default.nix (hunk 1 above)
Comment (gavinrogers) [reply pkgs/servers/holochain-go/default.nix]: Ok I'll remove the platforms metadata

**Comment by gavinrogers** [comment]
@GrahamcOfBorg build holochain-go

**Comment by xeji** [comment]
I'm afraid you lack the access rights to trigger the build.
@GrahamcOfBorg build holochain-go

**Comment by gavinrogers** [comment]
Ok it needs github.com/gxed/GoEndian, but only on linux? I don't fully understand why it's only complaining on some platforms. I should add it, but perhaps I should find out what is using it, and package that separately, with GoEndian as a dependency?

**Comment by zimbatm** [comment]
@GrahamcOfBorg build holochain-go

@gavinrogers make sure to use `--option sandbox true` when building on Linux. It's not enabled by default but is required by Hydra to make sure that all the dependencies are declared. It's possible that when testing without the sandboxing the go toolchain will fetch these from the network directly.

**Comment by zimbatm** [comment]
Also I recommend vendoring the dependencies using a tool like [godep](https://github.com/tools/godep). This will simplify the nix packaging as the deps.nix doesn't need to be generated anymore, and also allow anyone with go installed to easily run `go get github.com/holochain/holochain-proto` to build their own binary version.

**Comment by gavinrogers** [comment]
@zimbatm "Please use dep or another tool instead." in the README

**Comment by gavinrogers** [comment]
@grahamc @zimbatm @svanderburg @edolstra could I get a merge on this? we have a hackathon coming up this weekend and it would be a good demonstration of nixpkgs if participants could install nix, add the unstable channel and just ```nix-shell -p holochain-go```

**Pull request merged**